### ISSUE TITLE
Proposal: let scope additions

### ISSUE THREAD

**Issue by roemhildtg** [canjs/can-stache#645]
If I try to use let in a stache template multiple times on the same variable name, it results in an endless browser loop.

This currently does not work:
```html
{{let field="abc"}}
<p>{{field}}</p>
{{let field="def"}}
<div>{{field}}</div>
```

Perhaps something like this would allow us to get the most out of `let` scopes, yet also re-use variable names in certain cases:

```
{{#let field="abc"}}
<p>{{field}}</p>
{{/let}}<!-- field is no longer referenced -->
{{#let field="def"}}
<div>{{field}}</div>
{{/let}}
```

# Use case:

I have a set of field objects, each one has a stache partial.

```javascript
[{
    name: 'field1',
    template: stache(`<${tag} properties:from="field"
        object:from="dirtyObject"
        value:bind="dirtyObject[field.name]" 
        error:bind="validationErrors[field.name]"
        on:fieldchange="checkField(scope.arguments)"></${tag}>`);
}]
```

This partial references some variable names: `field`, `dirtyObject` each of which are available on different scopes. Using let scopes lets us do this. But it only lets us do it once.

## Alternatives

in my template, I can do something like this:

```
{{fields[0].template(field=fields[0])}} 
```

But this doesn't let us use let scopes either and instead creates a new context on top of the current one. Therefore `dirtyObject` and `object` are not available in the partial.

Perhaps calling a partial could use let scopes somehow as well.

**Comment by justinbmeyer** [comment]
Yeah, I planned to make this work.  Should be pretty straightforward. 

Hopefully `options.fn` only exists here: https://github.com/canjs/can-stache/blob/master/helpers/-let.js#L8 if `{{#` is used.  So you'd have something like:

```js
if( options.fn ) {
  return options.fn( options.scope.addLetContext( options.hash ) )
} else {
  // code that exists now
}
```

**Comment by roemhildtg** [comment]
Looks like `options.fn` is `noop` whenever its not in a block. So I checked for `isSection` instead. That appears to work.

**Comment by justinbmeyer** [comment]
This needs documentation to be officially closed.

**Comment by cherifGsoul** [comment]
Docs released as [v4.17.8](https://github.com/canjs/can-stache/releases/tag/v4.17.8)
